README
Tue Feb 16 13:48:04 2010

This program is an X11 Session Manager program.  It is started as the
last or only  command in a .xinitrc or .xsession script file.  It
provides a user definable menu of commands to be launched.  It also 
provides a text area that can be used for (electronic) note taking.  
The contents of the text area can be edited, saved, cleared, copied to
the X11 copy buffer, and printed.

It does the following on startup:
 1. Creates a pipe in /tmp that it reads text from.  This text is
    displayed on the Session Manager's text area. This pipe is also
    bound to stdout and stderr of the processes launched by from the
    user defined menu.
 2. Starts up the window manager.
 3. Optionally starts the Gnome Settings Daemon.
 4. Runs a session script which launches an initial set of programs.

I wrote this program as a replacement for the DECWindows session manager
that was included with the VAX/VMS DECWindows installation.  It is
written entirely in Tcl/Tk, making use of the standard packages,
BWidget, SNIT, and Img.

Building it should be trivial (Tcl/Tk (-devel packages), BWidget,
Tcllib, tetex w/ tetex-tex4ht, and doxygen need to be installed for
building from source):

./configure
make
make install

It uses tclkit technology to create a standalong native executable, so
once built, it is not necessary for Tcl/Tk to be installed on a target
system.  It is possible to built a 32-bit Linux executable on a 64-bit
machine (useful if you have 32-bit target systems).

To use the program, the user needs to create a file named
".tksessionmanagerrc" in his or her home directory.  This file is in
standard X11 resource format, containing lines association patterns to
values:

pattern: value

Where pattern is a standard X11 resource pattern and value is an
approprieate value.  These toplevel resources are used by the session
manager:

mainTitle (class MainTitle)
	Specifies the main title.  The default is "TK Session Manager".
mainGeometry (class MainGeometry)
	Specifies the size and placement of the session manager window.
	The default is to use the natural size and to center the window
	on the screen.
menuFilename (class MenuFilename)
	Specifies the name of the file containing the commands menu.
	The default is $HOME/tkSessionManager.menu.
printCommand (class PrintCommand)
	Specifies the command to use to print the contents of the
	session manager's text area.  Should be a command that can take
	a plain text stream on its stdin. Defaults to lp or lpr.
pipeName (class PipeName)
	Specifies the name of the pipe created in the /tmp directory.
	Text written to this pipe is displayed on the session manager's
	text area.  The default is ${USER}_TkSessionManager.
windowManager (class WindowManager)
	Specifies the path to the window manager program to start. 
	Defaults to /usr/bin/fvwm.
sessionScript (class SessionScript)
	Session startup script to run.  This stript contains the
	commands to start up the initial set of processes for the user's
	session. The default is $HOME/tkSessionManager.session.
gnomeSettingsDaemon (class GnomeSettingsDaemon)
	Flag to specify if the Gnome Settings Daemon should be started.
	This might be needed to allow theme settings for GTK+ 2
	programs. The default is yes.
gnomeScreensaver (class GnomeScreensaver)
	Flag to specify if the Gnome Screensaver should be allowed
	to run.  The Gnome Settings Daemon forks the Gnome Screensaver,
	which may not be desirable.  The default is no.

Here is an example .tksessionmanagerrc:

*MenuFilename: ~/mm.menu
*MainTitle: Robert's Menu
*MainGeometry: 590x315-20+110
*PrintCommand: a2ps -2
*background:  white
*foreground:  blue
*borderColor:  orange
*iconName:  Robert's Menu
*WindowManager: /usr/bin/fvwm
*SessionScript: ~/.x.login
*GnomeSettingsDaemon: yes
*GnomeScreensaver: no

Menu File
=========

The menu file consists of pairs of lines, the menu item text and the
command to run (which should be something suitable as an argument list
to the Tcl exec command).  Lines starting with a ! are comments ar
ignored. A casscade is introduced by using a '{' at the beginning of the
menu item text.  Lines are processed as menu items under the casscade
name until a lone '}' on a line by itself.  Casscades can occur under
casscades.  There is no set limit to the depth casscades can go.

Session Script
==============

The Session Script file is a normal script or program to run at startup.
Generally, it will be a series of commands that start up X11 programs. 
Each command should end in a '&', so that the programs are forked into
the background.  The script should be allowed to run to completion after
forking all of the programs.


ChangeLog
Fri Feb 19 21:03:15 2010

	1.0.0 Released.  This is the first public release.


Scripts/fileentry.tcl
#* 
#* ------------------------------------------------------------------
#* BWExtras.tcl -- Assorted extra composite widgets
#* ------------------------------------------------------------------
#* fileentry.tcl - File Entry Widget
#* Created by Robert Heller on Wed Feb 15 19:19:24 2006
#* ------------------------------------------------------------------
#* Modification History: $Log$
#* Modification History: Revision 1.2  2007/04/19 17:23:23  heller
#* Modification History: April 19 Lock Down
#* Modification History:
#* Modification History: Revision 1.1  2006/02/26 23:09:24  heller
#* Modification History: Lockdown for machine xfer
#* Modification History:
#* Modification History: Revision 1.1.1.1  2006/02/16 14:58:07  heller
#* Modification History: Imported sources
#* Modification History:
#* ------------------------------------------------------------------
#* Contents:
#* ------------------------------------------------------------------
#*  
#* Copyright (c) 2006, Robert Heller
#* All rights reserved.
#* 
#* Redistribution and use in source and binary forms, with or without
#* modification, are permitted provided that the following conditions are
#* met:
#* 
#*     * Redistributions of source code must retain the above copyright
#*       notice, this list of conditions and the following disclaimer.
#*     * Redistributions in binary form must reproduce the above copyright
#*       notice, this list of conditions and the following disclaimer in the
#*       documentation and/or other materials provided with the distribution.
#*     * Neither the name of the Deepwoods Software nor the names of its
#*       contributors may be used to endorse or promote products derived from
#*       this software without specific prior written permission.
#* 
#* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#* IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
#* TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
#* PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
#* OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
#* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
#* LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
#* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
#* THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
#* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
#* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#* 
#* 

#@Chapter:fileentry.tcl -- Labeled file entry megawidget.
#$Id: fileentry.tcl 761 2009-09-28 02:08:20Z heller $
# This is a specialized form of the LabelEntry widget intended for selecting
# file names.  A button is included to the right of the entry that pops up
# a file selection dialog.  Many of the resources from LabelFrame, Entry, and 
# Button are included in this widget.
#
# <option> -filebitmap The name of a bitmap to use for the button.  By default
#			an option folder image is used.
# <option> -fileimage The name of an image to use for the button.  By default
#			an option folder image is used.
# <option> -filedialog The type of file dialog to use.  Should be one of
#			open, save, or directory.  If open, tk_getOpenFile is
#			used; if save, tk_getSaveFile is used; and if directory,
#			tk_chooseDirectory is used.
# <option> -defaultextension This option is passed to tk_getOpenFile or 
#			tk_getSaveFile.
# <option> -filetypes This option is passed to tk_getOpenFile or tk_getSaveFile.
# <option> -title This option is passed to tk_getOpenFile or tk_getSaveFile.
# <option> -labeljustify From LabelFrame (-justify).
# <option> -labelwidth From LabelFrame (-width).
# <option> -labelanchor From LabelFrame (-anchor).
# <option> -labelheight From LabelFrame (-height).
# <option> -labelfont From LabelFrame (-font).
# <option> -labeltextvariable From LabelFrame (-textvariable).
# <option> -label From LabelFrame (-text).
# <option> -initialdir This option is passed to tk_getOpenFile or tk_getSaveFile.
# <option> -initialfile This option is passed to tk_getOpenFile or tk_getSaveFile.
# <option> -entryfg From Entry (-foreground).
# <option> -entrybg From Entry (-background).
# <option> -text From Entry.
# <option> -buttonfg From Button (-foreground).
# <option> -buttonbg From Button (-background).
# <option> -buttonactivebg From Button (-activebackground).
# <option> -buttonactivefg From Button (-activeforeground).
# <option> -buttondisabledfg From Button (-disabledforeground).
# <option> -buttonhighlightbg From Button (-highlightbackground).
# <option> -buttonhighlightcolor From Button (-highlightcolor).
# [index] FileEntry!widget

# ------------------------------------------------------------------------------
#  Index of commands:
#     - FileEntry::create
#     - FileEntry::configure
#     - FileEntry::cget
#     - FileEntry::bind
# ------------------------------------------------------------------------------

package require BWidget

namespace eval FileEntry {
# The namespace where this widget lives.
# [index] FileEntry!namespace

    proc use {} {}
    Widget::define FileEntry fileentry Entry LabelFrame Button
    Widget::declare FileEntry {
	{-filebitmap String "" 1}
	{-fileimage  String "" 1}
	{-filedialog Enum open 0 {open save directory}}
	{-defaultextension String "" 0}
	{-filetypes String {} 0}
	{-initialdir String {} 0}
	{-initialfile String {} 0}
	{-title String "" 0}
	{-modifycmd String "" 0}
    }

    Widget::bwinclude FileEntry LabelFrame .labf \
        remove {-relief -borderwidth -focus} \
        rename {-text -label} \
        prefix {label -justify -width -anchor -height -font -textvariable}

    Widget::bwinclude FileEntry Entry .e \
        remove {-fg -bg} \
        rename {-foreground -entryfg -background -entrybg}

    Widget::bwinclude FileEntry Button .b \
        remove {-anchor -bg -bitmap -borderwidth -bd -cursor -font
		-fg -highlightthickness -image -justify -padx -pady 
		-repeatdelay -repeatinterval -takefocus -text -textvariable 
		-wraplength -armcommand -command -default -disarmcommand 
		-height -helptext -helptype -helpvar -name -relief -state 
		-underline -width} \
	rename {-foreground -buttonfg -background -buttonbg
		-activebackground -buttonactivebg 
		-activeforeground -buttonactivefg
		-disabledforeground -buttondisabledfg
		-highlightbackground -buttonhighlightbg
		-highlightcolor -buttonhighlightcolor}

    Widget::addmap FileEntry "" :cmd {-background {}}

    Widget::syncoptions FileEntry Entry .e {-text {}}
    Widget::syncoptions FileEntry LabelFrame .labf {-label -text -underline {}}

    ::bind BwFileEntry <FocusIn> [list focus %W.labf]
    ::bind BwFileEntry <Destroy> [list FileEntry::_destroy %W]

    
}


# ------------------------------------------------------------------------------
#  Command FileEntry::create
# ------------------------------------------------------------------------------
proc FileEntry::create { path args } {
# Creation procedure
# <in> path -- The megawidget's path.
# <in> args -- Options for this widget.
# [index] FileEntry::create!procedure

    array set maps [list FileEntry {} :cmd {} .labf {} .e {} .b {}]
    array set maps [Widget::parseArgs FileEntry $args]

    eval [list frame $path] $maps(:cmd) -class FileEntry \
	    -relief flat -bd 0 -highlightthickness 0 -takefocus 0
    Widget::initFromODB FileEntry $path $maps(FileEntry)
    
	
    set labf  [eval [list LabelFrame::create $path.labf] $maps(.labf) \
                   [list -relief flat -borderwidth 0 -focus $path.e]]
    set subf  [LabelFrame::getframe $labf]
    set entry [eval [list Entry::create $path.e] $maps(.e)]
    set button [eval [list Button::create $path.b] $maps(.b)]
    set filebitmap "[Widget::getoption $path -filebitmap]"
    set fileimage  "[Widget::getoption $path -fileimage]"
    if {[string equal "$filebitmap" {}] &&
	[string equal "$fileimage"  {}]} {
      set fileimage [image create photo -file [file join $::BWIDGET::LIBRARY images openfold.gif]]
      $button configure -image "$fileimage"
    } elseif {![string equal "$filebitmap" {}]} {
      $button configure -bitmap $filebitmap
    }  elseif {![string equal "$fileimage"  {}]} {
      $button configure -image "$fileimage"
    }
    $button configure -command [list FileEntry::OpenFile $path]
    
    pack $entry -in $subf -fill both -expand yes -side left
    pack $button -in $subf -side right
    pack $labf  -fill both -expand yes

    bindtags $path [list $path BwFileEntry [winfo toplevel $path] all]

    return [Widget::create FileEntry $path]
}


# ------------------------------------------------------------------------------
#  Command FileEntry::configure
# ------------------------------------------------------------------------------
proc FileEntry::configure { path args } {
# Configuration procedure: configure one or more options for this widget.
# <in> path -- The path of the megawidget.
# <in> args -- Option value pairs.
# [index] FileEntry::configure!procedure

    return [Widget::configure $path $args]
}


# ------------------------------------------------------------------------------
#  Command FileEntry::cget
# ------------------------------------------------------------------------------
proc FileEntry::cget { path option } {
# Configuration option accessor procedure: access one option directly.
# <in> path -- The path of the megawidget.
# <in> option -- The option to access
# [index] FileEntry::cget!procedure

    return [Widget::cget $path $option]
}


# ------------------------------------------------------------------------------
#  Command FileEntry::bind
# ------------------------------------------------------------------------------
proc FileEntry::bind { path args } {
# Bind function.  Passthrough to the entry widget.
# <in> path -- The path of the megawidget.
# <in> args -- Bind arguments
# [index] FileEntry::bind!procedure

    return [eval ::bind [list $path.e] $args]
}


#------------------------------------------------------------------------------
#  Command FileEntry::_path_command
#------------------------------------------------------------------------------
proc FileEntry::_path_command { path cmd larg } {
# Path command for this megawidget.  Implements all of the megawidget commands.
# <in> path -- The path of the megawidget.
# <in> cmd -- The command name.
# <in> larg -- The command argument.
# [index] FileEntry::_path_command!procedure

    if { [string equal $cmd "configure"] ||
         [string equal $cmd "cget"] ||
         [string equal $cmd "bind"] } {
        return [eval [list FileEntry::$cmd $path] $larg]
    } else {
        return [eval [list $path.e:cmd $cmd] $larg]
    }
}


proc FileEntry::_destroy { path } {
# Destructor function.
# <in> path -- The path of the megawidget. 
# [index] FileEntry::_destroy!procedure

    Widget::destroy $path
}

#---------------------------------
# Bound to the button -- open a file select dialog
#---------------------------------

proc FileEntry::OpenFile { path } {
# Prodedure bound to the file open button.  Pops up a file selector dialog.
# <in> path -- The path of the megawidget.
# [index] FileEntry::OpenFile!procedure

  set dialogType [Widget::getoption $path -filedialog]
  set defaultextension [Widget::getoption $path -defaultextension]
  set filetypes [Widget::getoption $path -filetypes]
  set initialdir [Widget::getoption $path -initialdir]
  set initialfile [Widget::getoption $path -initialfile]
  set title [Widget::getoption $path -title]
  set modifycmd [Widget::getoption $path -modifycmd]
#  puts stderr "*** FileEntry::OpenFile: path = $path, dialogType = $dialogType, defaultextension = $defaultextension, filetypes = $filetypes, title = $title"

  set currentfile "[$path.e cget -text]"
  if {"$initialdir" eq "" || "$currentfile" ne ""} {
    if {"$dialogType" eq "directory"} {
      set initialdir "$currentfile"
    } else {
      set initialdir [file dirname "$currentfile"]
    }
  }
  if {"$initialfile" eq "" || "$currentfile" ne ""} {
    set initialfile "$currentfile"
  }
  switch $dialogType {
    open {
	set newfile [tk_getOpenFile \
			 -defaultextension "$defaultextension" \
			 -filetypes "$filetypes" \
			 -title "$title" \
			 -initialdir $initialdir \
			 -initialfile "$initialfile" \
			 -parent $path]
	if {![string equal "$newfile" {}]} {
	  $path.e configure -text "$newfile"
	  if {"$modifycmd" ne ""} {uplevel \#0 $modifycmd}
	}
    }
    save {
	set newfile [tk_getSaveFile \
			 -defaultextension "$defaultextension" \
			 -filetypes "$filetypes" \
			 -title "$title" \
			 -initialdir $initialdir \
			 -initialfile "$initialfile" \
			 -parent $path]
	if {![string equal "$newfile" {}]} {
	  $path.e configure -text "$newfile"
	  if {"$modifycmd" ne ""} {uplevel \#0 $modifycmd}
	}
    }
    directory {
	set newdirectory [tk_chooseDirectory \
		-initialdir "$initialdir" \
		-title "$title" \
		-parent $path]
	if {![string equal "$newdirectory" {}]} {
	  $path.e configure -text "$newdirectory"
	  if {"$modifycmd" ne ""} {uplevel \#0 $modifycmd}
	}
    }
  }

}

package provide BWFileEntry 1.0.0
